File: mobile/lib/core/providers/bkash_adapter.dart
import 'payment_models.dart';
import 'provider_adapter.dart';
import 'provider_id.dart';

/// Mock bKash adapter.
///
/// Phase B scope: LIVE for Tokenized Checkout merchant payments via the
/// backend; MOCK + dialer pass-through for P2P and Cash Out.
class BkashAdapter implements ProviderAdapter {
  @override
  ProviderId get id => ProviderId.bkash;

  @override
  Future<PaymentInitiation> initiate(PaymentRequest req) async {
    if (req.type == 'MERCHANT_PAYMENT') {
      return PaymentInitiation(
        providerTxnId: 'BKASH-M-${DateTime.now().millisecondsSinceEpoch}',
        redirectUrl: 'https://checkout.sandbox.bka.sh/mock?id=${req.idempotencyKey}',
        status: 'INITIATED',
        expiresAt: DateTime.now().add(const Duration(minutes: 10)).toIso8601String(),
      );
    }
    return launchDialerPassThrough(req);
  }

  @override
  Future<PaymentInitiation> launchDialerPassThrough(PaymentRequest req) async {
    final taka = req.amountMinorUnits ~/ 100;
    final ussd = '*247*1*${req.recipientPhone ?? ''}*$taka#';
    return PaymentInitiation(
      providerTxnId: 'BKASH-D-${DateTime.now().millisecondsSinceEpoch}',
      status: 'INITIATED',
      instructions: PaymentInstructions(method: 'DIALER_PASS_THROUGH', ussdString: ussd),
    );
  }

  @override
  Future<PaymentStatus> pollStatus(String providerTxnId) async => PaymentStatus(
        providerTxnId: providerTxnId,
        status: 'SUCCESS',
        updatedAt: DateTime.now().toIso8601String(),
      );

  @override
  Future<PaymentReceipt> fetchReceipt(String providerTxnId) async => PaymentReceipt(
        providerTxnId: providerTxnId,
        providerId: ProviderId.bkash,
        type: 'MERCHANT_PAYMENT',
        status: 'SUCCESS',
        amountMinorUnits: 0,
        currency: 'BDT',
        feeMinorUnits: 0,
        completedAt: DateTime.now().toIso8601String(),
      );
}

File: backend/src/modules/payments/providers/bkash.adapter.ts
import { randomUUID } from 'node:crypto';

import { mockInitiation, mockStatus } from './mock.helpers.js';
import type {
  PaymentInitiation,
  PaymentReceipt,
  PaymentRequest,
  PaymentStatus,
  ProviderAdapter,
} from './provider.interface.js';

/**
 * Mock bKash adapter.
 *
 * Phase B scope:
 *   - Merchant payments: LIVE via bKash Tokenized Checkout REST API.
 *     The production implementation goes here; for this scaffold it
 *     returns a mock redirect URL.
 *   - P2P / Cash Out: MOCK + dialer pass-through (USSD *247#).
 *
 * The live bKash implementation will replace the mock paths below in a
 * follow-up task. The interface and DTOs stay identical.
 */
export class BkashAdapter implements ProviderAdapter {
  readonly id = 'bkash' as const;

  async initiate(req: PaymentRequest): Promise<PaymentInitiation> {
    if (req.type === 'MERCHANT_PAYMENT') {
      return this.initiateMerchant(req);
    }
    return this.launchDialerPassThrough(req);
  }

  private initiateMerchant(req: PaymentRequest): Promise<PaymentInitiation> {
    // TODO(B3-live-providers): swap with real bKash tokenized checkout call.
    return Promise.resolve(
      mockInitiation(req, {
        providerTxnId: `BKASH-M-${randomUUID()}`,
        redirectUrl: `https://checkout.sandbox.bka.sh/v1.2.0-beta/payment/complete?mock=1&id=${req.idempotencyKey}`,
      }),
    );
  }

  launchDialerPassThrough(req: PaymentRequest): Promise<PaymentInitiation> {
    const ussd = `*247*1*${req.recipientPhone ?? ''}*${Math.floor(req.amountMinorUnits / 100)}#`;
    return Promise.resolve(
      mockInitiation(req, {
        providerTxnId: `BKASH-D-${randomUUID()}`,
        ussdString: ussd,
      }),
    );
  }

  pollStatus(providerTxnId: string): Promise<PaymentStatus> {
    return Promise.resolve(mockStatus(providerTxnId, 'SUCCESS'));
  }

  fetchReceipt(providerTxnId: string): Promise<PaymentReceipt> {
    return Promise.resolve({
      providerTxnId,
      providerId: 'bkash',
      type: 'MERCHANT_PAYMENT',
      status: 'SUCCESS',
      amountMinorUnits: 0,
      currency: 'BDT',
      feeMinorUnits: 0,
      completedAt: new Date().toISOString(),
      metadata: { mock: true },
    });
  }
}
